### PR TITLE
Don't imply that multiple drivers have errors

### PR DESCRIPTION
**Pull request by ann0see**
**Short description of changes**

If there's only one driver with an error message, there's no need to imply that multiple devices have errors. This change is done to simplify the error message.

The following is not exactly what my (fixed) code looks like.
error message

**Context: Fixes an issue?**

Idea from https://github.com/jamulussoftware/jamulus/discussions/2163#discussioncomment-1788094
**Does this change need documentation? What needs to be documented and how?**
No

**Status of this Pull Request**

Draft. It remains to test what happens if lNumDevs > 1.
**What is missing until this pull request can be merged?**
Untested since it's just a quick fix. Will test it by checking the artifacts.
## Checklist

- [x] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
- [ ] I tested my code and it does what I want
- [ ] My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#source-code-consistency) 
- [ ] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. 
- [ ] I've filled all the content above

### CODE REVIEW

**Comment by gilgongo** [comment]
Do you think we can remove the ambiguity in the first part of the message? The driver(s) in question my in fact be perfectly usable, it's just that there is a problem with them that may or may not be temporary. 

So if only one driver is erroring, just present this error message with an option to open the settings:

image

If multiple drivers are erroring, try to let the errors speak for themselves:

image

In general, long dialogue messages tend not to get read though. I find that had to accept, and don't really know why it is, but it's a very observable and common thing. So the other approach might be not to show the errors themselves and instead just have a call to action to open the settings? Depends on how helpful the messages are I guess:

image

**Comment by pljones** [comment]
I like the first two.  In the second, I'd add "If needed" at the start of the guidance.

I don't like the third one because it's far too vague to decide what to do.  "Maybe something is wrong with one of your sound devices but I'm not going to tell you which one or what the problem was."  Thanks....

**Comment by gilgongo** [comment]
@pljones Ah yes the last one doesn't indicate which card has the fail. OK that's no good then :-)

**Comment by pljones** [comment]
> BTW why "if needed"? Seems redundant. Or implies you may not want to fix the fact that Jamulus doesn't work for you?

Because you list drivers that failed.  If I'm not using those drivers, why would I want to correct them?  The
> You may be able to correct these in the driver settings

makes it sound like it _could_ **need** doing.  It's better to make it clear the user needs to decide what steps to take: we're not telling them to fix it - just that something went wrong that _could_ need fixing.

**Comment by gilgongo** [comment]
Oh I see - in "plural" errors case, "If needed... " would be clearer there but it still seems odd in the "singular" error case to me.

Agree about the first and last line prominence. Generally though, popups that spawn popups are seen as a bad thing as you can lose track or the parent or child if you accidentally click on something (and I think it's a bit of a mare for those using assistive tech). Also don't forget the ASIO settings is yet another "popop" in this context. But I guess if there's more than about three error messages it might be OK (the best solution would be an expandable panel within the dialogue but I assume that's not going to be possible).

> makes it sound like it should be looked at and may be fixable.

I disagree. The sense is that may or may not be issues to fix, and our lack of specificity clues this as well. Any other wording will be unacceptably long and complicated and we'll be back where we started.

File: src/soundbase.cpp [hunk 1]
```diff
@@ -166,22 +166,38 @@ QString CSoundBase::SetDev ( const QString strDevName )
...
 #ifdef _WIN32
             // to be able to access the ASIO driver setup for changing, e.g., the sample rate, we
             // offer the user under Windows that we open the driver setups of all registered
             // ASIO drivers
-            sErrorMessage += "<br/>" + tr ( "Do you want to open the ASIO driver setup to try changing your configuration to a working state?" );
+            sErrorMessage += "<br/>" + tr ( "You may be able to fix errors in the driver settings." );
+            // for Windows use a QMessageBox to show the error
+
+            QMessageBox mbWinErrorBox = QMessageBox::information ( nullptr, APP_NAME, sErrorMessage, QMessageBox::Cancel );
+            QPushButton*             btnASIOSettings = mbWinErrorBox.addButton ( tr ( "Open ASIO settings" ), QMessageBox::AcceptRole );
```
Comment (ann0see) [inline src/soundbase.cpp:198]: Sorry for being unsure about the syntax here.

File: src/soundbase.cpp [hunk 2]
```diff
@@ -166,30 +166,38 @@ QString CSoundBase::SetDev ( const QString strDevName )
         if ( !vsErrorList.isEmpty() )
         {
             // create error message with all details
-            QString sErrorMessage = "<b>" + QString ( tr ( "No usable %1 audio device found." ) ).arg ( strSystemDriverTechniqueName ) + "</b><br>";
+            QString sErrorMessage = "<b>";
```
Comment (pljones) [inline src/soundbase.cpp:169]: Just declare it here - the floating `"<b>"` makes the code hard to follow.

Comment (ann0see) [reply src/soundbase.cpp:169]: Ok. Still struggeling with the syntax though.

File: src/soundbase.cpp [hunk 3]
```diff
@@ -166,30 +166,38 @@ QString CSoundBase::SetDev ( const QString strDevName )
...
                 {
-                    sErrorMessage += "<li><b>" + GetDeviceName ( i ) + "</b>: " + vsErrorList[i] + "</li>";
+                    sErrorMessage += "<b>" + GetDeviceName ( i ) + "</b>: " + vsErrorList[i] + "<br>";
                 }
-                sErrorMessage += "</ul>";
             }
             else
             {
-                 sErrorMessage += "<b>" + GetDeviceName ( 0 ) + "</b>: " + vsErrorList[0] ;
+                sErrorMessage += tr ( "Couldn't initialize audio driver (using %1)</b><br>"
```
Comment (pljones) [inline src/soundbase.cpp:184]: Line 173:
> `"Sound driver errors (using %1) found:</b><br>"`

Line 184:
> `"Couldn't initialize audio driver (using %1)</b><br>"`

Use the same string in both places.  Ideally the less technical one.  The first doesn't need to end with ":" as it no longer introduces a list.

Maybe
> `"Could not find a sound driver (using %1) that could be used.</b><br>"`

(and do away entirely with `"Your sound device is not compatible in its current state."` as it would then be somewhat redundant).

File: src/soundbase.cpp [hunk 4]
```diff
@@ -166,30 +166,38 @@ QString CSoundBase::SetDev ( const QString strDevName )
...
+                sErrorMessage += "<b>" + GetDeviceName ( 0 ) + "</b>: " + vsErrorList[0];
             }
 
-
 #ifdef _WIN32
             // to be able to access the ASIO driver setup for changing, e.g., the sample rate, we
             // offer the user under Windows that we open the driver setups of all registered
             // ASIO drivers
-            sErrorMessage += "<br/>" + tr ( "You might be able to fix this error by changing settings of your ASIO driver. Do you want to open these settings now?" );
+            sErrorMessage += "<br/>" + tr ( "You may be able to fix errors in the driver settings." );
```
Comment (pljones) [inline src/soundbase.cpp:194]: This seems worthless.

File: src/soundbase.cpp [hunk 5]
```diff
@@ -166,30 +166,38 @@ QString CSoundBase::SetDev ( const QString strDevName )
         if ( !vsErrorList.isEmpty() )
```
Comment (pljones) [inline src/soundbase.cpp:166]: Lines 154 to 165 are irrelevant if this we enter this `if`.  They should be placed after the block.

File: src/soundbase.cpp [hunk 6]
```diff
@@ -166,30 +166,38 @@ QString CSoundBase::SetDev ( const QString strDevName )
...
+            sErrorMessage += "<br/>" + tr ( "You may be able to fix errors in the driver settings." );
+            // for Windows use a QMessageBox to show the error
+
+            QMessageBox::information mbWinErrorBox ( nullptr, APP_NAME, sErrorMessage, QMessageBox::Cancel );
+            QPushButton*             btnASIOSettings = mbWinErrorBox.addButton ( tr ( "Open ASIO settings" ), QMessageBox::AcceptRole );
 
-            if ( QMessageBox::Yes == QMessageBox::information ( nullptr, APP_NAME, sErrorMessage, QMessageBox::Yes | QMessageBox::No ) )
+            if ( mbWinErrorBox.clickedButton() == btnASIOSettings )
             {
                 LoadAndInitializeFirstValidDriver ( true );
```
Comment (pljones) [inline src/soundbase.cpp:202]: `vsErrorList = LoadAndInitializeFirstValidDriver ( true );` and then check the list size.  If it's zero (and I've not checked the code to make sure I'm right here...), `return strReturn` rather than continuing.

Comment (ann0see) [reply src/soundbase.cpp:202]: I think there's more work needed then just that. See the commit I mentioned: Probably it's in a deeper level.

**Comment by ann0see** [comment]
Ok. Maybe something like that would be ok:

image

File: src/soundbase.cpp [hunk 7]
```diff
@@ -166,24 +166,36 @@ QString CSoundBase::SetDev ( const QString strDevName )
...
-            if ( QMessageBox::Yes == QMessageBox::information ( nullptr, APP_NAME, sErrorMessage, QMessageBox::Yes | QMessageBox::No ) )
+            if ( qmASIOWarningBox.clickedButton() == btnASIOSettings )
             {
-                LoadAndInitializeFirstValidDriver ( true );
+                vsErrorList = LoadAndInitializeFirstValidDriver ( true );
+            }
+
+            if ( vsErrorList.isEmpty() ) {
+                return strReturn;
             }
```
Comment (ann0see) [inline src/soundbase.cpp:199]: As mentioned before, I think this is not enough.

File: src/soundbase.cpp [hunk 8]
```diff
@@ -166,24 +166,36 @@ QString CSoundBase::SetDev ( const QString strDevName )
...
-            sErrorMessage += "<br/>" + tr ( "Do you want to open the ASIO driver setup to try changing your configuration to a working state?" );
+            sErrorMessage += "<br>" + tr ( "You may be able to fix errors in the driver settings." );
+            // for Windows use a QMessageBox to show the error
+            QMessageBox qmASIOWarningBox;
+            QPushButton *btnASIOSettings = qmASIOWarningBox.addButton ( tr ( "Open ASIO settings" ), QMessageBox::AcceptRole );
+
+            qmASIOWarningBox.addButton ( QMessageBox::Cancel );
+            qmASIOWarningBox.setText ( sErrorMessage );
+            qmASIOWarningBox.setIcon ( QMessageBox::Warning );
+            qmASIOWarningBox.exec();
```
Comment (ann0see) [inline src/soundbase.cpp:190]: Probably this can be simplified somehow

Comment (pljones) [reply src/soundbase.cpp:190]: I think there's a constructor that allows most of those things to be set?

File: src/soundbase.cpp (hunk 7 above)
Comment (pljones) [reply src/soundbase.cpp:199]: It looks like it's what's needed and removes my concern at https://github.com/jamulussoftware/jamulus/pull/2168/files#r770944946

File: src/soundbase.cpp (hunk 8 above)
Comment (ann0see) [reply src/soundbase.cpp:190]: Yes, I think/thought so too. However I couldn't find it in the documentation (or the code gave an error). The problem was with adding a "custom" button.

File: src/soundbase.cpp (hunk 5 above)
Comment (ann0see) [reply src/soundbase.cpp:166]: Might the problem be that this code is run once at startup and also during runtime if someone changes the sound devices?

File: src/soundbase.cpp [hunk 9]
```diff
@@ -166,24 +166,35 @@ QString CSoundBase::SetDev ( const QString strDevName )
         if ( !vsErrorList.isEmpty() )
         {
             // create error message with all details
-            QString sErrorMessage = "<b>" + QString ( tr ( "No usable %1 audio device found." ) ).arg ( strSystemDriverTechniqueName ) +
-                                    "</b><br><br>" + tr ( "These are all the available drivers with error messages:" ) + "<ul>";
+            QString sErrorMessage = tr ( "<b>Could not find compatible sound driver (using %1).</b><br><br>" ).arg ( strSystemDriverTechniqueName );
```
Comment (gilgongo) [inline src/soundbase.cpp:169]: The "... (using ASIO)." seems a bit confusing here as it implies a driver might be found using something else. Is there a reason we're saying this?

Comment (pljones) [reply src/soundbase.cpp:169]: It won't necessarily say "ASIO" - it provides meaningful information for diagnosing a problem preventing the use of the program.

File: src/soundbase.cpp (hunk 5 above)
Comment (pljones) [reply src/soundbase.cpp:166]: Before line 197 was introduced, the variable set there wasn't referenced until after this if block.  However, now line 197 _is_ there, it makes sense.

File: src/soundbase.cpp (hunk 9 above)
Comment (gilgongo) [reply src/soundbase.cpp:169]: OK, but we have a button that says "Open ASIO settings", so is that button conditional on what `%1` is? This PR is to help people who aren't technical deal with the problem. If we qualify an error message in this way, they'd be excused for wondering why it's qualified, should they be using something else, etc.

Comment (pljones) [reply src/soundbase.cpp:169]: No -- it only appear on Windows.  (It should also only appear for ASIO -- that needs fixing.)

Comment (gilgongo) [reply src/soundbase.cpp:169]: I still think "Could not find compatible sound driver (using ASIO)" feels a bit odd. Like it's expecting them to switch to something that isn't ASIO where the problem might not arise.

Comment (pljones) [reply src/soundbase.cpp:169]: But that's entirely true.  Switching to Jamulus for Jack on Windows might solve the problem.

Comment (gilgongo) [reply src/soundbase.cpp:169]: It would be less mysterious to just make it the subject in that case and say, "Could not find a compatible ASIO sound driver." That is, unless we want also to say "Maybe you should try using [other system]." which I doubt.

File: src/soundbase.cpp [hunk 10]
```diff
@@ -166,24 +166,35 @@ QString CSoundBase::SetDev ( const QString strDevName )
...
 
             for ( int i = 0; i < lNumDevs; i++ )
             {
-                sErrorMessage += "<li><b>" + GetDeviceName ( i ) + "</b>: " + vsErrorList[i] + "</li>";
+                sErrorMessage += "<b>" + GetDeviceName ( i ) + "</b>: " + vsErrorList[i] + "<br><br>";
             }
-            sErrorMessage += "</ul>";
 
 #ifdef _WIN32
```
Comment (pljones) [inline src/soundbase.cpp:176]: Is this defined if we're running JACK on Windows?  If so, it's the wrong symbol.

File: src/soundbase.cpp (hunk 9 above)
Comment (pljones) [reply src/soundbase.cpp:169]: Maybe go back to `No usable %1 audio device found.` then?

Comment (gilgongo) [reply src/soundbase.cpp:169]: That reads better than a parenthesised suffix (I think). 

BTW re. "audio device", it's probably the subject of a different ticket, but for such an important thing we should keep to a standard term. The [style guide](https://jamulus.io/contribute/Style-and-Tone) says we should be using "sound hardware" here but "audio device" does feel a bit more conventional somehow.

**Comment by pljones** [comment]
Before approval, this PR should show behavour when:
- results for _all_ platforms (iOS, Android, MacOS, Linux, Window JACK and Windows ASIO)
- results when the final `throw CGenErr ( sErrorMessage );` occurs
- results when the final `return strReturn;` occurs
- okay to say "platform tested but could not achieve failure", though, I suppose...

This is needed to ensure it remains sane...

File: src/soundbase.cpp (hunk 10 above)
Comment (ann0see) [reply src/soundbase.cpp:176]: I think so. Similar to the ASIO Device setting button.

**Comment by ann0see** [comment]
I think it still throws an exception without returning. We‘d need some more debugging, but that would be out of scope of this PR.

File: src/soundbase.cpp (hunk 10 above)
Comment (pljones) [reply src/soundbase.cpp:176]: OK, probably needs a bug raised to fix it, then.

File: src/soundbase.cpp (hunk 9 above)
Comment (ann0see) [reply src/soundbase.cpp]: Ok. Fixed that.

> The whole flow is really bad, I think - huge block of platform-specific code at level that should be platform-independent.

I'm a bit unsure, but I remember there was a discussion about fixing this. Wasn't @npostavs involved?

Comment (npostavs) [reply src/soundbase.cpp]: I guess pljones' remark at https://github.com/jamulussoftware/jamulus/pull/1378#issuecomment-812866670 is similar.

**Comment by ann0see** [comment]
No longer sure if this is moving into the right direction. Probably the exception handling needs to be part of another PR.

File: src/soundbase.cpp [hunk 11]
```diff
@@ -166,24 +166,35 @@ QString CSoundBase::SetDev ( const QString strDevName )
...
 
-            if ( QMessageBox::Yes == QMessageBox::information ( nullptr, APP_NAME, sErrorMessage, QMessageBox::Yes | QMessageBox::No ) )
+            qmASIOWarningBox.addButton ( QMessageBox::Cancel );
+            qmASIOWarningBox.setText ( sErrorMessage );
+            qmASIOWarningBox.setIcon ( QMessageBox::Warning );
+            qmASIOWarningBox.exec();
+
+            if ( qmASIOWarningBox.clickedButton() == btnASIOSettings )
+            {
+                vsErrorList = LoadAndInitializeFirstValidDriver ( true );
```
Comment (ann0see) [inline src/soundbase.cpp:192]: I think we should revert this to the original state and do functional changes here later.
Related to https://github.com/jamulussoftware/jamulus/issues/2238

File: src/soundbase.cpp (hunk 10 above)
Comment (ann0see) [reply src/soundbase.cpp]: Ok. Fixed this now.

File: src/soundbase.cpp (hunk 11 above)
Comment (ann0see) [reply src/soundbase.cpp]: Done. This should be handled in another PR.

File: src/soundbase.cpp [hunk 12]
```diff
@@ -167,23 +167,32 @@ QString CSoundBase::SetDev ( const QString strDevName )
...
+
+            qmASIOWarningBox.addButton ( QMessageBox::Cancel );
+            qmASIOWarningBox.setText ( sErrorMessage );
+            qmASIOWarningBox.setIcon ( QMessageBox::Warning );
+            qmASIOWarningBox.exec();
+
+            if ( qmASIOWarningBox.clickedButton() == btnASIOSettings )
             {
+                // TODO: This or some related code aborts the app even if the driver is valid after user changes.
+                // This should be handled differently. For reference, please see https://github.com/jamulussoftware/jamulus/pull/2168
```
Comment (ann0see) [inline src/soundbase.cpp:195]: @pljones wrote a TODO comment here.

**Review by ann0see — COMMENTED** [review]
Rebased this PR.

**Comment by pljones** [comment]
Overall, far too much of the code appears to be trying to handle Windows-ASIO specific issues in `src/soundbase.cpp`, rather than `windows/sound.cpp`.  I'd rather _any_ improvements were focussed on removing ASIO code from `src/*`.

**Review by ann0see — COMMENTED** [review]
I think the text part should be split out into another PR (and the addition of the open ASIO Settings button overthought again)

**Comment by ann0see** [comment]
Closing in favour of https://github.com/jamulussoftware/jamulus/pull/2496 we should probably split this PR.
